### PR TITLE
Added a zero check for the parameter w.

### PR DESCRIPTION
**Pull request by Klavishnik**
In the function `static int bytepad:549` there is a division by the function parameter `w`.
If this parameter is set to zero, a division by zero will occur.

Therefore, a zero check for the parameter `w` was added at the beginning of the function.

Below is an example of the sanitizer output if the function `bytepad` is passed a parameter w of 0.

image

### CODE REVIEW

File: providers/implementations/macs/kmac_prov.c [hunk 1]
```diff
@@ -535,6 +535,9 @@ static int bytepad(unsigned char *out, size_t *out_len,
                    const unsigned char *in1, size_t in1_len,
                    const unsigned char *in2, size_t in2_len, size_t w)
 {
+    if (w == 0)
+        return 0;
+
```
Comment (tmshort) [inline providers/implementations/macs/kmac_prov.c:540]: Declarations must be first in the function.

Comment (tmshort) [reply providers/implementations/macs/kmac_prov.c:540]: The `if` statement is above the declarations that are on line 541. It needs to be below them.
Basically, C89 (which I still believe is the standard OpenSSL follows) does not allow mixing of declarations within code; all declarations must be at the start of a block. This change violates that.

**Comment by t8m** [comment]
This would be eligible for CLA: trivial. Please amend the git commit message body to include CLA: trivial on a separate line.

Or submit a regular CLA. https://www.openssl.org/policies/cla.html

**Comment by tmshort** [comment]
I was thinking it might be trivial, too.

**Comment by tom-cosgrove-arm** [comment]
@Klavishnik While you're updating the commit to add `CLA: trivial` on a line on its own, would you mind updating the commit message?

  `Added a zero check for the parameter w` is fine in the context of this function, but for someone just looking at commit messages it's a bit vague. Something like `Avoid diviide-by-zero in kmac_prov.c's bytepad()`?

File: providers/implementations/macs/kmac_prov.c [hunk 2]
```diff
@@ -535,6 +535,9 @@ static int bytepad(unsigned char *out, size_t *out_len,
                    const unsigned char *in1, size_t in1_len,
                    const unsigned char *in2, size_t in2_len, size_t w)
 {
+    if (w == 0)
```
Comment (tom-cosgrove-arm) [inline providers/implementations/macs/kmac_prov.c:538]: A few lines down there is

```
if (!ossl_assert(w <= 255))
    return 0;
```

Would it be a better fix to change the assert to `w > 0 && w <= 255` at move it to before `w` gets used?

**Comment by t8m** [comment]
My suggestion would be instead of fixing this in the bytepad() function, ensure that the parameter cannot be 0 at callers.

This is the block size of the hash which cannot really be 0 unless some internal error happens. Currently the code checks for EVP_MD_get_block_size() returning negative numbers, that should be just augmented to check for negative or zero return.

Doing what @tom-cosgrove-arm in addition to that would be also good, but not required.

**Comment by slontis** [comment]
How are you forcing it to get here?

The digest is set to either KECCAK-KMAC-128 or 256 internally..
EVP_MD_get_block_size() returns -1 if there is no digest.
So the only way I could see this happening is if the blocksize of the digest returned zero.

The point just after loading of the digest during kmac_fetch_new() is where it would then be incorrect.

**Comment by tom-cosgrove-arm** [comment]
> This is the block size of the hash which cannot really be 0 unless some internal error happens

Which is why I think adding to the assert (and moving that assert to a better place) is most appropriate

**Comment by t8m** [comment]
In theory the underlying MD implementation for the KMAC can come from a different provider which means it can return anything including 0. It would be wrong to use ossl_assert() instead of proper error in that case.

**Comment by Klavishnik** [comment]
What should we do about this issue?

**Comment by t8m** [comment]
> What should we do about this issue?

Could you please do what I suggested and change the line https://github.com/openssl/openssl/blob/bec6633b66c3506699b87d2cc226fbeb7ea3d8d8/providers/implementations/macs/kmac_prov.c#L292 to have `<= 0` there? You can keep your original change as is there as well.

**Comment by slontis** [comment]
@t8m this is not a sufficient change.. There are the 2 places that get the block_size.

Change lines 292 and 252 to <= 0
And move the assert line up.

File: providers/implementations/macs/kmac_prov.c [hunk 3]
```diff
@@ -538,7 +538,7 @@ static int bytepad(unsigned char *out, size_t *out_len,
     int len;
     unsigned char *p = out;
     int sz = w;
-
+
```
Comment (t8m) [inline providers/implementations/macs/kmac_prov.c:541]: You're adding end-of-line whitespace here. Please revert that.

**Review by t8m — APPROVED** [review]
Now you could also move the

```
    if (!ossl_assert(w <= 255))
        return 0;
```
before line 542 and modify the condition to `!ossl_assert(w > 0) || !ossl_assert(w <= 255)`

But it is optional.

**Review by paulidale — APPROVED** [review]

**Review by t-j-h — APPROVED** [review]

**Comment by t8m** [comment]
Squashed and merged to master, 3.1, and 3.0 branches. Thank you for your contribution.

**Pull request closed without merging**